Clipboard suggestions

codokie opened the pull request:
I made the primary clipboard appear in the suggestion strip view as a suggested word whenever neutral suggestions are to be shown like in a beginning of sentence (fixes #167).
Here is an example of the feature:
![example](https://github.com/Helium314/HeliBoard/assets/151087174/521054eb-cdc6-406d-b393-fb550b65ab53)


In android 7.0+, the API allows detection of sensitive clipboard content, and when that happens the content is redacted like this:
![password](https://github.com/Helium314/HeliBoard/assets/151087174/cd1b3443-4cbc-4f2b-99ae-3b26d4446a5c)

Suggestions will disappear if they are picked or 3 minutes have passed.
There is also new setting in Suggestions settings that allows to toggle the suggestions.
There will be a suggestion for number input fields only if the clipboard content is a number.

P.S. You can long press the clipboard suggestion and then press the X icon to dismiss it

RHJihan commented:
Nice work. ❤
One question: Is there a time limit for how long old clipboard content will be suggested?
The clipboard icon from the toolbar can be used for consistency, perhaps?

Helium314 commented:
As usual, the PR is already 2 days old but I didn't even have time to go through the code...

I haven't tested, but "whenever neutral suggestions are to be shown" might not be quite the right thing.
HeliBoard might start showing suggestions right away when opening an input field, but when this is right after copying, the clipboard suggestion might still be wanted. Though the clipboard suggestion could be added to the list of normal suggestions in such a case.
On the other hand seeing the clipboard suggestion often for a long time might be seen as annoying / unnecessary.
I'm thinking about when it should not be shown any more. I'll ask in the issue to reach a few more users.

And maybe the icon should be the normal clipboard icon instead? I can imagine the icon looking a bit out of place depending on the theme / colors.

RHJihan commented:
> I'm thinking about when it should not be shown any more. I'll ask in the issue to reach a few more users.

My opinion is up to 90 seconds right after copying any text. Samsung keyboard does this for about 120 seconds.

RHJihan commented:
> Suggestions will disappear if they are picked or if they are older than max history retention time.

I am a little bit confused. What if someone toggles off the clipboard history (the default is 10 min) or sets the retention time to a bigger one? Or set that to _No limit_? Is it convenient to match the history retention time or to set some specific time, i.e., 90 seconds or 120 seconds?

> There will be a suggestion for number input fields only if the clipboard content is a number.

It's impressive. Loved it.

> The unicode clipboard icon has been replaced with the default clipboard icon.

It's looking good.

codokie commented:
> I am a little bit confused. What if someone toggles off the clipboard history (the default is 10 min) or sets the retention time to a bigger one? Or set that to No limit? Is it convenient to match the history retention time or to set some specific time, i.e., 90 seconds or 120 seconds?

I can easily set it to a constant time but personally feel that even 120 seconds is too short. If I copied something and something has caught my attention for 2 minutes (may happen frequently), I still want to be suggested the copied content afterwards.

If retention is set to "No limit", then the content will be suggested until the user has picked the suggestion or deleted it (by long pressing and pressing the bin button, very easy).
It will also stop appearing after the primary clipboard has been cleared (either manually or hourly in Android 13+).

It is possible to add a slider for setting the time but I haven't seen a keyboard who actually has such a setting.

Glad you liked the other additions :)

codokie commented:
@RHJihan I've checked this again and I think you may be right. So I've set the suggestion to expire every 2 minutes.

If you have the time, could you please review the code or test this PR?

Also, I've noticed that in many cases when there are no (clipboard) suggestions and we are not in the middle of composing text, the suggestions are empty.
I think it would make more sense to show the main toolbar instead of the "neutral" suggestions. That's how it works in GBoard etc.

So in [ba3f49b](https://github.com/Helium314/HeliBoard/pull/647/commits/ba3f49b261b1df8fc58b22f62ebd3818da953f93) I've also added that functionality, but if that's problematic it can be moved to a separate PR (will cause a merge conflict though).

RHJihan commented:
> Actually it is inconsistent, should I revert it back?

I am confused here, too. But I think clipboard suggestions shouldn't be very long. Anyone needing clipboard content after a specific time can always have it on the clipboard history view.

I tested the PR. Here are my thoughts:
If the history retention time is less than 120 seconds, I think clipboard suggestions should match that time.

If I copy text when another keyboard is active, after 120 seconds, it still shows as a clipboard suggestion on this PR. I am not sure if it is related to this PR. Maybe it's start counting the time after loading the keyboard. But it's not an important thing after all.

I am not that level of coder. So, I am giving my experience and perspective as a user and/or tester. 



> I think it would make more sense to show the main toolbar instead of the "neutral" suggestions. That's how it works in GBoard etc.

I think it's cool.

codokie commented:
> I think clipboard suggestions shouldn't be very long.

If others won't claim that 2 min is too short then I'll leave it at that.
> If the history retention time is less than 120 seconds, I think clipboard suggestions should match that time.

I have added a setting called "Clear primary clipboard" in #611 (still open) to clear the primary clipboard too after history retention time.
With that setting enabled, it should match this behavior. Currently the history retention time only affects the internal clipboard and the clipboard suggestions use the primary one.
> If I copy text when another keyboard is active, after 120 seconds, it still shows as a clipboard suggestion on this PR.

I think it is because the variables that keep information (timestamp, etc.) about the last primary clipboard entry are destroyed when the keyboard is changed. To fix this I can move those variables to the companion object.
> I am giving my experience and perspective as a user and/or tester

That's exactly what I was looking for. Thank you for the feedback!

RHJihan commented:
It is perfect now for me.

> To fix this I can move those variables to the companion object.

After a quick try, this still happening. Force stop the keyboard app and enable it again will suggest old clipboard content as suggestion.

RHJihan commented:
> long press the clipboard suggestion and then press the bin icon to remove it

Is Samsung Keyboard's way of deleting suggestions better? 

![Screenshot_20240410_143317_Keep Notes](https://github.com/Helium314/HeliBoard/assets/31377578/10f381b7-6554-48f0-ab9a-71664f962ace)

RHJihan commented:
> By the way, I would appreciate some help in making the suggestion background distinct from the suggestion strip view background like in other apps. I tried modifying it programmatically but it seems to affect only the pressed state of the TextView.

As I said earlier, I am not that level of coder. So I cannot help you with this. @Helium314 might help you.

> I could adjust the bin button to just dismiss the suggestion without affecting the primary/internal clipboard, but I think the icon may convey a message that the entry should be deleted.

What about both? X button to dismiss the clipboard suggestion, but it can be accessed later on the clipboard history if a user would like to retain history. And the long press bin button to delete the clipboard entry.

I'm probably asking for much change. I am sorry for that.

codokie commented:
That's ok thanks for the genuine feedback. I'll see if I can add such button

codokie commented:
On a second thought, such a button would have to replace the pinned toolbar keys to ensure the suggestion strip view is not cluttered/confusing.
The pinned keys may still be useful for the user whenever there is a clipboard suggestion

RHJihan commented:
> On a second thought, such a button would have to replace the pinned toolbar keys to ensure the suggestion strip view is not cluttered/confusing.

Is it possible to place that button after any pinned toolbar button or on the left side right after the toolbar button?

codokie commented:
There are keyboards like Gboard who do not offer a way to dismiss the clipboard suggestion. Not saying that this is good, but it offers a more minimal/clean user interface.

I think having 2 keys, one for dismissing suggestion and another for dismissing it and deleting the entry, is a bit redundant and currently not offered by any mainstream keyboard.

It may make more sense to replace the bin icon with a dismiss icon when long pressing a clipboard suggestion, which just dismisses the suggestion without altering the primary/internal clipboard. Do you agree with me on this?

In the aforementioned PR #611 I added the "Clear history" as toolbar key, so you would still be able to clear your clipboard with just a tap.

RHJihan commented:
If you think having 2 keys is redundant, I agree with that. I like your simple approach. 

For dismissing the clipboard suggestion, the x button is more discoverable compared to the long press dismissal from my point of view. What do you think about this?

codokie commented:
@RHJihan  You are definitely right that it is not visible but I can't say objectively if it's bad that way because it is relatively intuitive to discover features by long pressing. Maybe it should be added to the "Hidden Functionality" readme?

So I replaced the bin icon with 'X' icon when long pressing a clipboard suggestion.
The touch zone is now bigger and covers the whole text view to ensure the clipboard is not pasted by accident when trying to dismiss the suggestion.

I hope you don't mind but I increased the retention limit to 3 minutes, still not sure of the right value to make this feature useful while being not obstructive..

RHJihan commented:
I am totally delighted with the way it is now, and I am waiting for this PR to be merged. 

> Maybe it should be added to the "Hidden Functionality" readme?

Yes, I can mention this to the [Readme](https://github.com/Helium314/HeliBoard/blob/main/README.md) and [Hidden features](https://github.com/Helium314/HeliBoard/blob/846c25a0a340871690133cc6520186a6d13d525b/app/src/main/res/values/strings.xml#L748).

codokie commented:
@Helium314 I cleaned some of the code and removed many changes not directly related to the clipboard suggestions from it (opened 2 separate PRs). Now it should be a lot easier to review it.
I also made it possible to show clipboard suggestions when "Show correction suggestions" is turned off.

RHJihan commented:
When the clipboard content is an image, the clipboard suggestion starts acting weirdly. 
I am guessing it's related to how HeliBoard handles clipboard images, as it does not support images. 

![Screenshot_20240414_122619_Keep Notes](https://github.com/Helium314/HeliBoard/assets/31377578/329130cb-54a9-4307-b55b-d1104ba42a1f)

codokie commented:
Thanks for the heads up, Good catch!
From my check the upstream Heliboard also exhibits this behavior when an image is copied to the clipboard, you just don't notice it until the history view is opened. If the image is high-res, it may even make the keyboard unresponsive.
I added a check to make sure we coerce the clipboard data to text only if it is text.
When you have the time please let me know if that helped

RHJihan commented:
Thanks. Now it's fixed.

codokie commented:
@Helium314 I think this PR is ready for beta testing, would be grateful if it could be reviewed when you've got the time for it. Thanks

RHJihan commented:
What do you think about this visual change?

Current:
![2](https://github.com/Helium314/HeliBoard/assets/31377578/5c2e7bd1-cacb-4a9d-9bdb-d312b1f98aff)

Proposed:
![1](https://github.com/Helium314/HeliBoard/assets/31377578/df93940c-4477-44f6-8a94-a274ea7e3f19)

BlackyHawky commented:
Very nice. Could you show a screenshot with light theme ?

RHJihan commented:
![1](https://github.com/Helium314/HeliBoard/assets/31377578/9af6990d-8df8-45c6-bac3-d636491b1ba4)

codokie commented:
I agree it looks great :+1: 
I tried to apply a similar background but failed to do so without breaking the color state list of the middle/clipboard suggestion.
Currently the clipboard suggestion shares the same TextView with a regular suggestion, so changing the background needs to be done on the fly in the SuggestionStripLayoutHelper class

Helium314 commented:
> @Helium314 I think this PR is ready for beta testing, would be grateful if it could be reviewed when you've got the time for it. Thanks

It will be a while until I'm through with this as there are many changes for something seemingly simple, but at least I have some time now.

For now I'll just go into some details I notice, and try getting an overview of what is going on later.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/suggestions/SuggestionStripView.java:
```diff
@@ -417,10 +420,10 @@ private boolean onLongClickSuggestion(final TextView wordView) {
         if (showIcon) {
-            final Drawable icon = mBinIcon;
+            final Drawable icon = mSuggestedWords.isClipboardSuggestion() ? mCloseIcon : mBinIcon;
             Settings.getInstance().getCurrent().mColors.setColor(icon, ColorType.REMOVE_SUGGESTION_ICON);
-            int w = icon.getIntrinsicWidth();
-            int h = icon.getIntrinsicWidth();
+            int w = wordView.getWidth();
+            int h = wordView.getHeight();
```
Does this still allow selecting a word once the bin icon is showing?

codokie replied:
Selecting a word once the bin icon is shown will cause the suggestion to be deleted.
I made this change because it seemed too easy to accidentally pick the (clipboard) suggestion when trying to delete it.

codokie replied:
In [d4083a3](https://github.com/Helium314/HeliBoard/pull/647/commits/d4083a386d87c8cc98f8a6f8b6448af05451a75c) I reverted this change as I figured out what was causing the suggestions to be accidentally picked.
OnLongClick() would not return true if the bin icon had not been touched, so the click event was being consumed by OnClick()

Helium314 commented on app/src/main/java/helium314/keyboard/latin/suggestions/SuggestionStripLayoutHelper.java:
Why are changes here actually necessary? It seems to me you're just showing a suggestion with an icon, which should already work.

codokie replied:
Here is a summary of the changes made to the `SuggestionStripLayoutHelper` class:
1) Redact sensitive clipboard content (replace with a mask of asterisks)
2) Truncate clipboard content longer than 15 characters so that it would not take too much screen estate
3) Set the layout width of the clipboard suggestion's TextView to wrap around the text (otherwise a clipboard suggestion will be picked when an "empty" space on the suggestion strip view is touched)
4) Add background color span to the text of the clipboard suggestion (it's less pretty than a proper background drawable, but I had no luck adding one that doesn't break the color state list of the TextView)
5) Add a clipboard icon to the left of the suggestion. Note that in the `SuggestionStripView` class, only the dismiss icon is added

Helium314 commented on app/src/main/res/layout/suggestions_strip.xml:
```diff
@@ -52,6 +52,7 @@
             android:layout_height="match_parent"
             android:layout_weight="1"
             android:hapticFeedbackEnabled="false"
+            android:gravity="center_horizontal"
```
What is the purpose / effect of this?

codokie replied:
This change of gravity is to ensure that a clipboard suggestion (which is not wide enough to fill the entire suggestion strip view) will be centered as opposed to aligned to the left (or to the right if an RTL subtype is selected)

Helium314 commented on app/src/main/res/values/strings.xml:
```diff
@@ -128,6 +128,10 @@
     <string name="bigram_prediction">Next-word suggestions</string>
     <!-- Description for "next word suggestion" option. This displays suggestions even when there is no input, based on the previous word. -->
     <string name="bigram_prediction_summary">Use the previous word in making suggestions</string>
+    <!-- Option to enable the suggestion of clipboard content. -->
+    <string name="suggest_clipboard_content">Suggest clipboard content</string>
+    <!-- Description for the "suggest clipboard content" option. This makes the primary clipboard content visible in the suggestion strip view. -->
+    <string name="suggest_clipboard_content_summary">Show the content of the primary clipboard as a suggestion</string>
```
Users will not know what a "primary clipboard" is. Maybe call it system clipboard, or just clipboard.
It should also mention that only recently copied items will be in the suggestions.

codokie replied:
How about `Show recently copied clipboard content as a suggestion`?

Helium314 replied:
Yes, that sounds good

Helium314 commented:
Btw since your PRs often touch the same regions of code, I decided to work on the ones related to clipboard from oldest to newest (this being the oldest).

Helium314 commented:
I wonder, is there a way to put the clipboard suggestion like an inline suggestion view?
It seems more suitable to me, as the clipboard suggestion is similar to inline autofill suggestion as in being a differently styled suggestion that is shown alone (without any other suggestions).

codokie commented:
I have to agree with you that the inline suggestions styling seems like a good fit for the clipboard suggestions, but I don't know yet if we can just 'steal' the style.
Code-wise the inline suggestions use a specialized API to communicate with the autofill app, it may seem out of place to whoever reads the code if we use some of it for a different purpose (if that's even possible).

I think the style @RHJihan suggested looks elegant as well, and it shouldn't require an overhaul to implement it.
@Helium314 if you have an idea how to programmatically change the color state list every time there is a suggestion, that could really help.
[This](https://github.com/Helium314/HeliBoard/blob/22abdb5d0857d1903154e967ccfe5e80dcaa45cc/app/src/main/java/helium314/keyboard/latin/suggestions/SuggestionStripLayoutHelper.java#L378) is the place where I believe the style change should be done. For some reason when I change the background color of the center view, only the pressed state is altered..

Helium314 commented:
> I have to agree with you that the inline suggestions styling seems like a good fit for the clipboard suggestions, but I don't know yet if we can just 'steal' the style.

Right, I hadn't noticed that the style is actually coupled to inline suggestions.

> Code-wise the inline suggestions use a specialized API to communicate with the autofill app, it may seem out of place to whoever reads the code if we use some of it for a different purpose (if that's even possible).

It should be easy to generalize the inline suggestion part in `SuggestionStripView`, after all it's just setting a View and keeping track of the visibility.
My idea was to use `setInlineSuggestionsView` to set some customized clipboard view like
```java
    public View createClipView(String clipText) {
        TextView view = new TextView(getContext());
        view.setOnClickListener((v) -> mListener.onClipboardSuggestionPicked(clipText));
        // add some styling
        return view;
    }
```
and `onClipboardSuggestionPicked` calls `LatinIME.onTextInput`.
The view also has a handler, so you could e.g. trigger removal after a certain time using `view.postDelayed`.

codokie commented:
I see, it does seem like that should work well for the clipboard suggestions, but the TextView still needs to be styled, and that's also my current gripe with the current implementation, particularly with the background drawable

codokie commented:
Ok so I think I managed to make it styled similarly to inline suggestions, thanks for the tip to make the clipboard suggestions a separate textview. Here is how it looks now:
![example](https://github.com/Helium314/HeliBoard/assets/151087174/521054eb-cdc6-406d-b393-fb550b65ab53)

Helium314 commented:
Since you still don't use `setInlineSuggestionsView` and didn't provide information on why, I had a deeper look, and I think it should be possible (and preferable, as per contribution guidelines).

The idea when clipboard suggestion could be shown:
First `LatinIME` checks whether clipboard suggestions are wanted, and if yes calls something like `mClipboardHistoryManager.createClipboardSuggestionView(editorInfo)` (analogous to `InlineAutofillUtils.createView`).
If there is suitable clipboard content, `ClipboardHistoryManager` returns a non-null view with the listeners already in place, and `LatinIME` calls mSuggestionStripView.setInlineSuggestionsView (as mentioned previously, this could be renamed).

Is there anything fundamentally wrong with this approach?

codokie commented:
It sounds doable and may reduce code from SuggestionStripView, but it would litter ClipboardHistoryManager, so I don't know about that..
Unlike inline suggestions, we need to keep reference of the clipboard suggestion view to programmatically change it inside the removeSuggestion(), onLongClickSuggestion() and clear() functions

Helium314 commented:
> it would litter ClipboardHistoryManager,

If you think ClipboardHistoryManager is going to grow too much, you could have the view creation happen in a utility file.

The re-use of the inline suggestion view should also reduce the need for clipboard-suggestion specific code in LatinIME.

> Also creating a new textview each time would be less efficient than just keeping reference to it since only the text needs to be altered

Possibly, depending on how often it's really created. If this is an issue, a prepared view can be cached, and updated before use.

> but there would be a need for an indication that the view is indeed a clipboard suggestion

Why is this? I don't have a full overview over the PR right now.
In any case it should not be an issue, the tag can be set to "clipboard" or something like that.

codokie commented:
> The re-use of the inline suggestion view should also reduce the need for clipboard-suggestion specific code in LatinIME.

But I did not use inline suggestion view for creating the clipboard suggestion view which is a TextView mostly defined in the xml.
I only used the background of the inline suggestion.

Sorry but I don't think it's worthwhile to use the inline view for this, if it's at all possible.

I understand wanting the reduce the code footprint but I think it makes sense to set the touch listener in the suggestion strip view and not in a utility file

Helium314 commented:
Seriously, those parts of the contribution guidelines have been there since OpenBoard times. And they don't exist just because someone likes writing guidelines.
For me the whole process is also very frustrating, as I have to discuss in order to convince you to follow these guidelines, and not just in this PR.

I really would like to spend my time on code instead of discussions, but this kind of adding yet another thing for almost the same functionality is making the app harder to maintain. And I understand you might not be concerned about maintainability ("_won't affect the appearance or usability of the feature_").
But _I need to_ maintain this app, and in some places it's already a nightmare where changes are very prone to have unintendend effects. I'm not exactly innocent in this respect (specifically looking at the keyboard parsing, which I tried to improve several times already for this reason), but there is a difference between this sort of issue arising for historic reasons and when it's plainly in sight and can be avoided with relatively low effort.
Re-using existing mechanism consistently contributes to maintainability of the code, because when making a change (much) later I don't have to figure out / remember why there are two different ways of showing a single "special" suggestion, and when doing changes I only have to be careful about not breaking one functionality instead of two.

I'm trying to improve the current situation, and I've had "_Internal cleanup (a lot of over-complicated and convoluted code)_" in the [to do section](https://github.com/Helium314/HeliBoard?tab=readme-ov-file#to-do) for a while now. Currently it feels like I have to not just (slowly) work on this, but also have to do discussions over and over again that such things do not find their way into the app via PRs.

I have "_still it would be good if you announced that you are working on it, so we can discuss how changes are best implemented_" in the [contribution guidelines](https://github.com/Helium314/HeliBoard/blob/main/CONTRIBUTING.md#guidelines), and I think discussing on how to implement larger changes _before_ most of the code is written is a good way to avoid me asking to completely change your code.
Could be in the issue, or by starting a draft PR of a (possibly) non-functional version sketching your idea.

I'm sorry if this comment comes across as overly criticising, so I want to explicitly mention that I definitely _do_ value your contributions (and I'm not happy that I don't have time to keep up with your pace of PRs).

codokie commented:
This PR is for adding clipboard suggestions, not abstracting the creation of views in the suggestion strip view.
I've tried before to use the existing mechanism of the inline suggestions but could not get it to work.
Clipboard suggestion is a TextView and its content is not generated by the autofill API.
If you want you can merge the PR as-is or implement the "relatively low effort" change you think is required.
I'm sorry but I don't time or will to work on this anymore

Helium314 commented:
I adjusted the PR to use the (now renamed) `setInlineSuggestionsView`, and create the view in `ClipboardHistoryManager`.
Behavior is same, or very close to, except for:
* always showing the dismiss icon (I assume users may want to dismiss the suggestion relatively often and should have an easy way to do it)
* suggestion is not disappearing automatically after 3 min (I think there is no real benefit from this)

Further I removed some crash sources (using methods / resources not available on devices with the current minApi).

RHJihan commented:
I haven't tested this merged PR yet. But I just wanted to know:

> suggestion is not disappearing automatically after 3 min (I think there is no real benefit from this)

So, will the clipboard suggestion appear every time (if the clipboard has something copied already) I try to type something unless I dismiss it manually?

Helium314 commented:
It will not appear if the clip is more than 3 min old. But it will not disappear on its own if you just wait and stare at the keyboard for 3 minutes.

codokie commented:
@Helium314 Thanks for finalizing this pull request. I look forward to test this feature in the upcoming beta

Helium314 commented:
I hope the implementation is ok for you. I tried to keep behavior close to your version when moving most of the stuff into that one view.
(Will also work on your other PRs, but that will take some time. Especially because I'll be barely available for the next 3 weeks.)
